spec/services/update_wiki_namespace_stats_spec.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe UpdateWikiNamespaceStats do
+  let(:course) { create(:course, start: Date.new(2022, 8, 1), end: Date.new(2022, 8, 2)) }
+  let(:wiki) { Wiki.get_or_create(language: 'en', project: 'wikibooks') }
+  let(:namespace) { 102 }
+  let(:user1) { create(:user, username: 'Jamzze') } # user with cookbook edits
+  let(:user2) { create(:user, username: 'FieldMarine') } # user with other namespace edits
+
+  before do
+    stub_wiki_validation
+    course.campaigns << Campaign.first
+    course.wikis << wiki
+    JoinCourse.new(course: course, user: user1, role: 0)
+    JoinCourse.new(course: course, user: user2, role: 0)
+    allow_any_instance_of(Course).to receive(:tracked_namespaces).and_return([0, 102])
+    VCR.use_cassette 'course_update' do
+      UpdateCourseStats.new(course)
+    end
+    described_class.new(course, wiki, namespace)
+  end
+
+  it 'adds articles with only tracked namespaces to article_courses' do
+    # namespaces of all the fetched revisions of course users
+    fetched_namespaces = course.revisions.joins(:article).distinct.pluck('articles.namespace')
+    # namespaces of article_courses
+    article_namespaces = course.articles.distinct.pluck(:namespace)
+
+    expect(fetched_namespaces).to include(0, 2, 3, 102)
+    expect(article_namespaces).to include(0, 102)
+    expect(article_namespaces).not_to include(2, 3)
+  end
+
+  it 'updates course_stat record with appropriate wiki-namespace key' do
+    expect(course.course_stat).not_to be_nil
+
+    stats = course.course_stat.stats_hash
+    expect(stats).to have_key('en.wikibooks.org-namespace-102')
+  end
+
+  it 'updates wiki-namespace stats with appropriate keys' do
+    stats = course.course_stat.stats_hash['en.wikibooks.org-namespace-102']
+
+    expect(stats).to have_key(:edited_count)
+    expect(stats).to have_key(:new_count)
+    expect(stats).to have_key(:revision_count)
+    expect(stats).to have_key(:user_count)
+    expect(stats).to have_key(:word_count)
+    expect(stats).to have_key(:reference_count)
+    expect(stats).to have_key(:view_count)
```

Comment:
What values do these have? Does this test exercise the generation of non-zero values for all of these stats? (I'm guessing no for the view count, and yes for the most of the rest?)

Reply:
Actually, I haven't checked for the values.
So, should I get the count for each attribute and write a test for it?

Reply:
Also, I wasn't sure if I should include the test where it checks for each attribute in key. Should I keep it?

Reply:
Yes. We might find that such values can vary a little bit over time (if any of the edits get deleted at some point) but for now, I think it's better to have a little extra test focus on verifying that these stats are actually being counted the way we think they should be.

Reply:
> Also, I wasn't sure if I should include the test where it checks for each attribute in key. Should I keep it?

Yes, I think that's fine. It makes it a little more explicit what shape we expect this data to take.

Reply:
@ragesoss, have updated the test.